caput/profile.py

```diff
@@ -163,3 +166,216 @@ def __exit__(self, *args, **kwargs):
+        cpu_percent :  float
+            Process CPU utilization since `start` was called as percentage. Can be >100 if multiple threads run on
+            different cores. See `PSUtil.cpu_count` for available cores.
+        disk_io : `dict`
+            dict version of `psutil.io_counters` (on Linux) or `psutil.disk_io_counters` (on MacOS).
+            Difference since `start` was called.
+        memory : str
+            Difference of memory in use by this process since `start` was called. If negative,
+            less memory is in use now.
+
+        Raises
+        ------
+        RuntimeError
+            If stop was called before start.
+        """
+        stop_time = time.time()
+
+        # Get all stats at the same time
+        with self.oneshot():
+            cpu_times = self.cpu_times()
+            cpu_percent = self.cpu_percent()
+            memory = self.memory_full_info().uss
+            if psutil.MACOS:
+                disk_io = psutil.disk_io_counters()
+            else:
+                disk_io = self.io_counters()
+
+        if label not in self._start_cpu_times:
+            raise RuntimeError(
+                f"PSUtilProfiler.stop was called before start for '{label}'."
+            )
+
+        # Construct results
+        self._usage[label] = {}
+
+        CPU_Times = collections.namedtuple("CPUtimes", list(cpu_times._fields))
+        cpu_times = CPU_Times(*np.subtract(cpu_times, self._start_cpu_times.pop(label)))
+        DiskIO = collections.namedtuple("DiskIO", list(disk_io._fields))
+        disk_io = DiskIO(*np.subtract(disk_io, self._start_disk_io.pop(label)))
+        memory = memory - self._start_memory.pop(label)
+
+        self._usage[label]["cpu_times"] = cpu_times._asdict()
+        self._usage[label]["cpu_percent"] = cpu_percent
+        self._usage[label]["disk_io"] = disk_io._asdict()
+
+        def bytes2human(num):
+            for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
+                if abs(num) < 1024.0:
+                    return f"{num:3.1f}{unit}B"
+                num /= 1024.0
+            return f"{num:.1f}YiB"
+
+        memory = bytes2human(memory)
+
+        self._usage[label]["memory"] = memory
+
+        time_s = stop_time - self._start_time.pop(label)
+
+        self._usage[label]["time_s"] = time_s
+
+        if time_s < 0.1 and self._logger:
+            self._logger.info(
+                f"{label} ran for {time_s:.4f} < 0.1s, results might be inaccurate.\n"
+            )
+
+        if self._logger:
+            self._logger.info(
+                f"{label} ran for {time_s:.4f}s\n"
+                f"---------------------------------------------------------------------------------------------"
+                f"\n{cpu_times}\n"
+                f"Average CPU load: {cpu_percent}\n"
+                f"{disk_io}\n"
+                f"Change in (uss) memory: {memory}\n"
```

Comment:
I think it's useful to report the total memory used here, possibly both for the process and systemwide, in addition to the delta.